### PR TITLE
Fixes #11444 - Correctly count hosts in domain

### PR DESCRIPTION
**Pull request by tbrisker**
Only hosts that have a primary nic assigned to a domain should be
counted in the domain host count. Previous implementation was flawed in
that it only updated the counter in certain occasions and not all,
leading sometimes to incorrect host counts.
I have implemented the counter updates in `after_commit` to reduce
chance of deadlocks.

### CODE REVIEW

**Comment by domcleal** [comment]
Please note that this PR needs rebasing.

File: app/models/nic/base.rb [hunk 1]
```diff
@@ -223,6 +220,36 @@ def mac_uniqueness
...
+          domain.increment!('total_hosts') if domain_id.present?
+        else #changed from primary to non-primary
+          if domain_changed
+            Domain.find(previous_changes['domain_id'][0]).decrement!('total_hosts') if previous_changes['domain_id'][0].present?
+          else
+            domain.decrement!('total_hosts') if domain_id.present?
+          end
+        end
+      else
+        return unless domain_changed && primary?
```
Comment (domcleal) [inline app/models/nic/base.rb]: `elsif primary? && domain_changed` above?

File: app/models/nic/base.rb [hunk 2]
```diff
@@ -223,6 +220,36 @@ def mac_uniqueness
...
+
+    def update_domain_counters_on_update
+      primary_changed = previous_changes.include? 'primary'
+      domain_changed  = previous_changes.include? 'domain_id'
+      if primary_changed
+        if primary? #changed from non-primary to primary
+          domain.increment!('total_hosts') if domain_id.present?
+        else #changed from primary to non-primary
+          if domain_changed
+            Domain.find(previous_changes['domain_id'][0]).decrement!('total_hosts') if previous_changes['domain_id'][0].present?
```
Comment (domcleal) [inline app/models/nic/base.rb]: Does this need to be `Domain.unscoped.find`, to bypass taxonomies in the default scope?

(ditto below)

**Comment by domcleal** [comment]
Thanks, it looks okay but the amount of custom counting code is a little worrying.  Would you mind adding some more test coverage, particularly for the non-primary interfaces, primary to non-primary and vice versa cases?

**Comment by tbrisker** [comment]
@domcleal I addressed your inline comments and added a fair bit of testing. I agree this is a lot of custom counting code. I'm starting to reconsider my stance on using [counter_culture](https://github.com/magnusvk/counter_culture) for the counters, both because of this and because of the various deadlock issues popping up around counters.

File: app/models/domain.rb [hunk 3]
```diff
@@ -30,7 +30,7 @@ class Domain < ActiveRecord::Base
   validates :fullname, :uniqueness => true, :allow_blank => true, :allow_nil => true
 
   scoped_search :on => [:name, :fullname], :complete_value => true
-  scoped_search :on => :hosts_count
+  scoped_search :on => :total_hosts
```
Comment (domcleal) [inline app/models/domain.rb]: This should be aliased back to the original name.

File: lib/tasks/fix_cached_counters.rake [hunk 4]
```diff
@@ -1,10 +1,15 @@
 desc 'Fix cached counters by reseting them to the correct count, in case they got corrupted somehow'
 task :fix_cached_counters => :environment do
   puts "Correcting cached counters: (this may take a few minutes)"
-  [ Architecture, Environment, Operatingsystem, Domain, Realm].each do |cl|
+  [ Architecture, Environment, Operatingsystem, Realm].each do |cl|
     cl.unscoped.all.each{|el| cl.reset_counters(el.id, :hosts, :hostgroups)}
     puts "#{cl} corrected"
   end
+  Domain.all.each do |el|
+    Domain.reset_counters(el.id, :hostgroups)
```
Comment (domcleal) [inline lib/tasks/fix_cached_counters.rake:9]: hosts?

Comment (domcleal) [inline lib/tasks/fix_cached_counters.rake:9]: Oh nevermind, I get it now.  I didn't even know this had a hostgroups count.  Why!?

**Comment by domcleal** [comment]
Thanks, the test coverage is better now.  The only comment is the one about keeping the search keyword.

The gem you suggested does appear to fix a few of the problems that we've encountered.  I'm happy to get these tests and fixes in now though to enable refactoring later.

File: lib/tasks/fix_cached_counters.rake (hunk 4 above)
Comment (tbrisker) [inline lib/tasks/fix_cached_counters.rake:9]: Actually I have no idea, and I think this isn't the only counter that is useless. I think it can be removed, but I would suggest that it be done as a different PR.

**Comment by tbrisker** [comment]
@domcleal I added the alias. I think we should look into counter_culture again after the rails 4 upgrade if we still run into these problems, as the counter_cache handling in rails 4 is a bit better.

**Comment by domcleal** [comment]
Merged as a152a1b, thanks @tbrisker.

**Comment by tbrisker** [comment]
Thanks @domcleal, I have opened two issues as per our discussion:
[#12484](http://projects.theforeman.org/issues/12484) - Remove useless counters
[#12485](http://projects.theforeman.org/issues/12485) - Consider counter_culture gem
